File: src/main/java/com/rokucraft/RokuBot/config/serializers/ButtonSerializer.java
package com.rokucraft.RokuBot.config.serializers;

import net.dv8tion.jda.api.entities.Emoji;
import net.dv8tion.jda.api.interactions.components.Button;
import net.dv8tion.jda.api.interactions.components.ButtonStyle;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.spongepowered.configurate.ConfigurationNode;
import org.spongepowered.configurate.serialize.SerializationException;
import org.spongepowered.configurate.serialize.TypeSerializer;

import java.lang.reflect.Type;

public class ButtonSerializer implements TypeSerializer<Button> {

    public static final ButtonSerializer INSTANCE = new ButtonSerializer();

    @Override
    public Button deserialize(Type type, ConfigurationNode node) throws SerializationException {
        Emoji emoji = null;
        if (!node.node("emoji").virtual()) {
            emoji = Emoji.fromMarkdown(node.node("emoji").getString());
        }
        return Button.of(
                node.node("style").get(ButtonStyle.class, ButtonStyle.LINK),
                node.node("id-or-url").getString(),
                node.node("label").getString(),
                emoji
        );
    }

    @Override
    public void serialize(Type type, @Nullable Button obj, ConfigurationNode node) throws SerializationException {
        if (obj == null) {
            node.raw(null);
            return;
        }
        node.node("style").set(obj.getStyle());
        if (obj.getStyle() == ButtonStyle.LINK) {
            node.node("id-or-url").set(obj.getUrl());
        } else {
            node.node("id-or-url").set(obj.getId());
        }
        node.node("label").set(obj.getLabel());
        if (obj.getEmoji() != null) {
            node.node("emoji").set(obj.getEmoji().getAsMention());
        }
    }
}

File: src/main/kotlin/com/rokucraft/rokubot/config/serializers/ButtonSerializer.kt
package com.rokucraft.rokubot.config.serializers

import net.dv8tion.jda.api.entities.emoji.Emoji
import net.dv8tion.jda.api.interactions.components.buttons.Button
import net.dv8tion.jda.api.interactions.components.buttons.ButtonStyle
import org.spongepowered.configurate.ConfigurationNode
import org.spongepowered.configurate.kotlin.extensions.get
import org.spongepowered.configurate.serialize.SerializationException
import org.spongepowered.configurate.serialize.TypeSerializer
import java.lang.reflect.Type

object ButtonSerializer : TypeSerializer<Button> {
    @Throws(SerializationException::class)
    override fun deserialize(type: Type, node: ConfigurationNode): Button {
        var emoji: Emoji? = null
        val emojiCode: String? = node.node("emoji").get()
        if (!emojiCode.isNullOrEmpty()) {
            emoji = Emoji.fromFormatted(emojiCode)
        }
        val idOrUrl: String = node.node("id-or-url").get()
            ?: throw SerializationException("A value is required for this field")
        return Button.of(
            node.node("style").get(ButtonStyle::class, ButtonStyle.LINK),
            idOrUrl,
            node.node("label").get(),
            emoji
        )
    }

    @Throws(SerializationException::class)
    override fun serialize(type: Type, obj: Button?, node: ConfigurationNode) {
        if (obj == null) {
            node.raw(null)
            return
        }
        node.node("style").set(obj.style)
        node.node("id-or-url").set(if (obj.style == ButtonStyle.LINK) obj.url else obj.id)
        node.node("label").set(obj.label)
        node.node("emoji").set(obj.emoji?.formatted)
    }
}
